metadata-ingestion/src/datahub/ingestion/source/schema_inference/object.py

```diff
@@ -0,0 +1,181 @@
+from collections import Counter
+from typing import Any
+from typing import Counter as CounterType
+from typing import Dict, Iterable, Tuple, Union
+
+from mypy_extensions import TypedDict
+
+
+class BasicSchemaDescription(TypedDict):
+    types: CounterType[type]  # field types and times seen
+    count: int  # times the field was seen
+
+
+class SchemaDescription(BasicSchemaDescription):
+    delimited_name: str  # collapsed field name
+    # we use 'mixed' to denote mixed types, so we need a str here
+    type: Union[type, str]  # collapsed type
+    nullable: bool  # if field is ever missing
+
+
+def is_nullable_doc(doc: Dict[str, Any], field_path: Tuple) -> bool:
+    """
+    Check if a nested field is nullable in a document from a collection.
+
+    Parameters
+    ----------
+        doc:
+            document to check nullability for
+        field_path:
+            path to nested field to check, ex. ('first_field', 'nested_child', '2nd_nested_child')
+    """
+
+    field = field_path[0]
+
+    # if field is inside
+    if field in doc:
+
+        value = doc[field]
+
+        if value is None:
+            return True
+
+        # if no fields left, must be non-nullable
+        if len(field_path) == 1:
+            return False
+
+        # otherwise, keep checking the nested fields
+        remaining_fields = field_path[1:]
+
+        # if dictionary, check additional level of nesting
+        if isinstance(value, dict):
+            return is_nullable_doc(doc[field], remaining_fields)
+
+        # if list, check if any member is missing field
+        if isinstance(value, list):
+
+            # count empty lists of nested objects as nullable
+            if len(value) == 0:
+                return True
+
+            return any(is_nullable_doc(x, remaining_fields) for x in doc[field])
+
+        # any other types to check?
+        # raise ValueError("Nested type not 'list' or 'dict' encountered")
+        return True
+
+    return True
+
+
+def is_nullable_collection(
+    collection: Iterable[Dict[str, Any]], field_path: Tuple
+) -> bool:
+    """
+    Check if a nested field is nullable in a collection.
+
+    Parameters
+    ----------
+        collection:
+            collection to check nullability for
+        field_path:
+            path to nested field to check, ex. ('first_field', 'nested_child', '2nd_nested_child')
+    """
+
+    return any(is_nullable_doc(doc, field_path) for doc in collection)
```

Comment:
I think this may be incorrect. The `collection` is a dictionary & `doc` will just be the key here. Is that the intention?